# UNIVERSIDAD NACIONAL DEL OESTE
## Escuela de Informática
### PROGRAMACIÓN CON OBJETOS I
#### Profesores: Mag. Ing. Pablo Pandolfo / Lic. Leandro Politano
#### Segundo Examen Parcial - Noviembre 2025

* **ALUMNO**:
* **DNI**:

1. Crear una clase llamada **Cancion** con los siguientes atributos: titulo, artista, duracion_minutos, reproducciones y genero. Debe incluir:
* **Constructor**
* Método **\_\_str__()** que devuelva la representación textual del objeto con el siguiente formato: "Flowers - Miley Cyrus (3.2 min, 2.500.000 reproducciones)"
* Método **es_popular()** que devuelva True si tiene más de 5.000.000 reproducciones.

In [38]:
class Cancion:

    def __init__(self, titulo, artista, duracion_minutos, reproducciones, genero):
        self.__titulo = titulo
        self.__artista = artista
        self.__duracion_minutos = duracion_minutos
        self.__reproducciones = reproducciones
        self.__genero = genero

    def __str__(self):
        return f'{self.__titulo} - {self.__artista} ({self.__duracion_minutos} min, {self.__reproducciones} reproducciones)'

    def __repr__(self):
        return self.__str__()
    
    def es_popular(self):
        return self.__reproducciones > 5000000

    def get_genero(self):
        return self.__genero

2. Implementar una clase llamada **Playlist** con un método **cargar_playlist(ruta_archivo)** que lea de un archivo con extensión csv que contiene canciones y almacene en una lista objetos **Canción**. Cada línea del archivo tiene la siguiente estructura:

| titulo    | artista     | duracion_minutos | reproducciones | genero    |
| --        | --          | --               | --             | --        |
| Flowers   | Miley Cyrus | 3.20             | 2500000        | Pop       |
| Despacito | Luis Fonsi  | 4.45             | 8000000        | Reggaeton |
| Unholy    | Sam Smith   | 3.15             | N/A            | Pop       |

Debe manejar las siguientes excepciones:
* **FileNotFoundError**: mostrar "Archivo no encontrado".
* **ValueError**: si alguna duración o cantidad de reproducciones no se puede convertir, mostrar "No es un número"
* Cualquier otra excepción debe capturarse y mostar "Error!!".

In [39]:
import csv
class Playlist:

    def __init__(self):
        self.__canciones = []

    def cargar_playlist(self, ruta_archivo):
        try:
            with open(ruta_archivo, "r", encoding="utf-8") as f:
                datos = csv.DictReader(f, delimiter=';')
                for reg in datos:
                    self.__canciones.append(Cancion(reg["titulo"], reg["artista"], float(reg["duracion_minutos"]), int(reg["reproducciones"]), reg["genero"]))
        except FileNotFoundError:
            print('Archivo no encontrado')
        except ValueError:
            print('No es un número')
        except Exception as e:
            print(e)

    def canciones_por_genero(self):
        mapa = {}
        for cancion in self.__canciones:
            if cancion.get_genero() not in mapa:
                mapa[cancion.get_genero()] = []
            mapa[cancion.get_genero()].append(cancion)
        return mapa

In [40]:
p = Playlist()
p.cargar_playlist('data/playlist.csv')
p.canciones_por_genero()

{'Pop': [Flowers - Miley Cyrus (3.2 min, 2500000 reproducciones),
  As It Was - Harry Styles (2.55 min, 6000000 reproducciones),
  Vampire - Olivia Rodrigo (3.45 min, 2000000 reproducciones),
  Dance The Night - Dua Lipa (3.35 min, 3 reproducciones)],
 'Reggaeton': [Despacito - Luis Fonsi (4.45 min, 8000000 reproducciones),
  Titi Me Preguntó - Bad Bunny (4.0 min, 9500000 reproducciones)],
 'R&B': [Kill Bill - SZA (3.25 min, 7000000 reproducciones),
  Snooze - SZA (3.2 min, 5500000 reproducciones)],
 'Bachata': [La Bachata - Manuel Turizo (3.05 min, 5000000 reproducciones)]}

3. Agregar a la clase Playlist el método **canciones_por_genero()** que devuelva una estructura donde: 
* **clave** sea el nombre del género
* **valor** sea una lista de canciones de ese género

4. V/F
* Un decorador no puede recibir argumentos adicionales.
* El método readlines() devuelve una lista con todas las líneas del archivo.
* Si un archivo no existe y se abre con el modo "r", Python lo crea automáticamente.
* Un diccionario en Python no puede tener claves repetidas.
* La instrucción raise se usa para capturar una excepción.

FVFVF

## Rúbricas

|     |                                                  |
| --  | --                                               |
| 1.0 | Clase Canción                                    |
| 1.0 | Lectura de archivo                               |
| 1.0 | Append de lista                                  |
| 1.0 | Excepciones                                      |
| 1.0 | Iteracion de lista                               |
| 2.5 | Generacion de diccionario                        |
| 2.5 | V/F                                              |

## Puntaje

* 6   -> 4
* 6.5 -> 4
* 7   -> 5
* 7.5 -> 5
* 8   -> 6
* 8.5 -> 7
* 9   -> 8
* 9.5 -> 9
* 10  -> 10